Nixpkgs Implementation (and Docker Improvement)

alexandre-lavoie opened the pull request:
*Fixes #1882*

# Summary

This pull request implements/improves useful tools to make CorsixTH more cross-compatible. Issue #1882 was made to discuss this further!

# Docker

- Fixes the Dockerfile, which used to be stopped because it asked for user input.
- Added a Docker-Compose file to be able to build using only `docker-compose up`.

# Nixpkgs

These commits implement almost all available features of CorsixTH in a `default.nix` file. The file is basically ready for Nixpkgs (https://github.com/NixOS/nixpkgs). It could be interesting to check further if there is a need for more CorsixTH tools part of the package. It could also be interesting to use the package for CI/CD.

- Adds all build "scripts/tools" in the `*.nix` file to get the game running.
- Includes a bunch of customization (audioSupport, movieSupport, etc) - which is essentially the same as the CMake configs.

TheCycoONE reviewed (approved):
Looks good to me. Thanks.

alexandre-lavoie commented on Dockerfile.build:
```diff
@@ -3,6 +3,23 @@ FROM ubuntu:latest
+    # Base Tools
+    build-essential cmake git doxygen \
+    # LUA
+    liblua5.2-0 liblua5.2-dev liblua5.3-0 liblua5.3-0-dbg liblua5.3-dev \
```
Given you find it weird aswell, I will remove 5.2 from the build and test (I'm guessing maintainers didn't notice with the one-line install)

alexandre-lavoie commented:
I removed the Lua 5.2 packages lingering in the Dockerfile.

I also cleaned up the Nixpkg to be more modular and added the 3 current "apps" (Main, LevelEdit, and AnimView).

I'm not 100% sure what output binary would be appropriate - I decided to name all files as: `corsix-th-*`.

If these changes are approved, I'll try to pull on `Nixos/nixpkgs` and we can merge when everything is okay!

TheCycoONE commented on AnimView/CMakeLists.txt:
```diff
@@ -57,7 +57,12 @@ else()
 endif()
 
-target_link_libraries(AnimView RncLib)
+if(NOT RncLib)
```
I think you are trying to use the cmake for AnimView in isolation? It's not meant to be used that way. You can cmake the root project with all components and then target individual components with make/ninja.

TheCycoONE commented on Dockerfile.build:
```diff
@@ -3,6 +3,23 @@ FROM ubuntu:latest
+    # Font
+    libfreetype6-dev
+
+CMD mkdir -p build && cd build && cmake -DUSE_SOURCE_DATADIRS=ON .. && make
```
Needs final newline

TheCycoONE commented on default.nix:
```diff
@@ -0,0 +1,19 @@
+    corsixth = callPackage ./nix/corsixth.nix { inherit baseName; inherit audioSupport; inherit movieSupport; inherit freetypeSupport; inherit buildDocs; inherit enableUnitTests; };
+    animView = if buildAnimView then callPackage ./nix/animView.nix { inherit baseName; } else null;
+    levelEdit = if buildLevelEdit then callPackage ./nix/levelEdit.nix { inherit baseName; } else null;
+})
```
Likewise

TheCycoONE commented on nix/animView.nix:
```diff
@@ -0,0 +1,54 @@
+{ baseName ? "corsix-th", stdenv, lib, cmake, wxGTK30 }:
```
Please don't do it this way. cmake the root, then make AnimView to build your target.

alexandre-lavoie commented:
> I think you are trying to use the cmake for AnimView in isolation? It's not meant to be used that way. You can cmake the root project with all components and then target individual components with make/ninja.

> Please don't do it this way. cmake the root, then make AnimView to build your target.

I had it this way initially (was building both in a single package). I'm wondering why it is the preferred way? Going through root forces me to have the SDL and Lua packages despite not using them for the AnimView.

TheCycoONE commented:
I'm open to a -DBUILD_CORSIXTH=OFF option being added to CMake, which would be simple enough to implement. It hasn't really come up that someone would want one of the ancillary programs without CorsixTH.

alexandre-lavoie commented:
> I'm open to a -DBUILD_CORSIXTH=OFF option being added to CMake, which would be simple enough to implement. It hasn't really come up that someone would want one of the ancillary programs without CorsixTH.

Agreed that it would be a weird config. I think it could be interesting to have for this type of package manager. I'll commit the config and we can see if it makes sense - else I can always just include the AnimView part of the main `Corsixth` package.

lewri commented on CMakeLists.txt:
```diff
@@ -126,8 +127,10 @@ message("Building common libraries")
 add_subdirectory("libs")
 
 # We always build CorsixTH otherwise we would miss the generated header
```
This comment is no longer correct

TheCycoONE replied:
I wish I knew what header the line was referring to.

TheCycoONE commented on nix/corsixth.nix:
```diff
@@ -0,0 +1,89 @@
+    version = "0.64";
+
+    lua = {
+        env = lua5_3.withPackages(ps: with ps; [ lpeg luafilesystem luasocket ]);
```
luasocket is currently only used for the update check. If you want to depend on the package manager you can leave it out.

alexandre-lavoie replied:
I was considering to remove it, just was not sure if it could cause issues down the line. All I know is we need to update through the package manager.

TheCycoONE commented on docker-compose.yml:
```diff
@@ -0,0 +1,10 @@
+
+services:
+  corsixth:
+    container_name: corsixth
```
corsix-th

TheCycoONE commented:
You're still missing a couple trailing newlines. Is it common to add nixpkg files upstream? It's not something I've noticed.

alexandre-lavoie commented:
> You're still missing a couple trailing newlines. Is it common to add nixpkg files upstream? It's not something I've noticed.

Usually people keep a `shell.nix` around to have the tools needed to build the repo (see https://github.com/OpenRCT2/OpenRCT2/blob/develop/shell.nix ). The structure is essentially a copy of what we would upload on nix package manager (see https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/openrct2/default.nix ). The way I have it allows for both `nix-build` and `nix-shell` to work. I am not 100% sure what is the standard regarding this, but I personally find it convenient to have it this way!

TheCycoONE commented on .gitignore:
```diff
@@ -101,3 +101,7 @@ CorsixTH/CorsixTH.sln
+
+# Nix
+*.so
+/result*
```
Final missing newline

alexandre-lavoie commented:
Fixed EOF newline pointed out. Let me know if I missed something else!

I'm looking into the procedure to get this official on nixpkgs. After abit of testing, I should be set!

alexandre-lavoie commented:
Would it be possible to merge before publishing to Nixpkgs? I need the changes made to the CMake to be able to build remote. (Everything seems to be OK, just getting issues because of the AnimView).

TheCycoONE commented:
Are you going to open a new PR with just the changes intended to merge to this project?

alexandre-lavoie commented:
Sorry for delays! I'm not 100% certain what is the best approach. I think keeping track of the changes that were made in this PR could be good to have. I will commit a removal of the nix files (for bloating/duplicating concerns). If you prefer me just PRing the CMakeFile changes I am open to do so aswell!

TheCycoONE commented:
Yep, going to squash it, but that looks ok.
